libvips/conversion/tilecache.c

```diff
@@ -602,156 +670,59 @@ static int
 
-	VIPS_DEBUG_MSG_RED(
-		"vips_tile_cache_gen: "
+	VIPS_DEBUG_MSG_RED("vips_tile_cache_gen: "
 		"left = %d, top = %d, width = %d, height = %d\n",
 		r->left, r->top, r->width, r->height);
 
 	/* Ref all the tiles we will need.
 	 */
-	work = vips_tile_cache_ref(cache, r);
+	if (!(work = vips_tile_cache_ref(cache, r))) {
+		g_mutex_unlock(cache->lock);
 
-	while (work) {
-		/* Search for data tiles: easy, we can just paste those in.
-		 */
-		for (;;) {
-			for (p = work; p; p = p->next) {
-				tile = (VipsTile *) p->data;
-
-				if (tile->state == VIPS_TILE_STATE_DATA)
-					break;
-			}
-
-			if (!p)
-				break;
-
-			VIPS_DEBUG_MSG_RED(
-				"vips_tile_cache_gen: pasting %p\n",
-				tile);
-
-			vips_tile_paste(tile, out_region);
-
-			/* We're done with this tile.
-			 */
-			work = g_slist_remove(work, tile);
-			vips_tile_unref(tile);
-		}
+		return -1;
+	}
 
-		/* Calculate the first PEND tile we find on the work list. We
-		 * don't calculate all PEND tiles since after the first, more
-		 * DATA tiles might heve been made available by other threads
-		 * and we want to get them out of the way as soon as we can.
-		 */
-		for (p = work; p; p = p->next) {
-			tile = (VipsTile *) p->data;
-
-			if (tile->state == VIPS_TILE_STATE_PEND) {
-				tile->state = VIPS_TILE_STATE_CALC;
-
-				VIPS_DEBUG_MSG_RED(
-					"vips_tile_cache_gen: calc of %p\n",
-					tile);
-
-				/* In threaded mode, we let other threads run
-				 * while we calc this tile. In non-threaded
-				 * mode, we keep the lock and make 'em wait.
-				 */
-				if (cache->threaded)
-					g_mutex_unlock(cache->lock);
-
-				/* Don't compute if we've seen an error
-				 * previously.
-				 */
-				if (!result)
-					result = vips_region_prepare_to(in,
-						tile->region,
-						&tile->pos,
-						tile->pos.left, tile->pos.top);
-
-				if (cache->threaded) {
-					VIPS_GATE_START("vips_tile_cache_gen: wait2");
-
-					g_mutex_lock(cache->lock);
-
-					VIPS_GATE_STOP("vips_tile_cache_gen: wait2");
-				}
-
-				/* If there was an error calculating this
-				 * tile, black it out and terminate
-				 * calculation. We have to stop so we can
-				 * support things like --fail on jpegload.
-				 *
-				 * Don't return early, we'd deadlock.
-				 */
-				if (result) {
-					VIPS_DEBUG_MSG_RED(
-						"vips_tile_cache_gen: error on tile %p\n",
-						tile);
-
-					g_warning(_("error in tile %d x %d"),
-						tile->pos.left, tile->pos.top);
-
-					vips_region_black(tile->region);
-
-					*stop = TRUE;
-				}
-
-				tile->state = VIPS_TILE_STATE_DATA;
-
-				/* Let everyone know there's a new DATA tile.
-				 * They need to all check their work lists.
-				 */
-				g_cond_broadcast(cache->new_tile);
-
-				break;
-			}
+	/* Loop until all our tiles are ready, or there's an error of some sort,
+	 * perhaps a loader hits a bad tile.
+	 */
+	GSList *todo;
+	todo = g_slist_copy(work);
+	while (todo) {
+		VipsTile *tile = (VipsTile *) todo->data;
+		todo = g_slist_remove(todo, tile);
```

Comment:
I think we could avoid the (possible expensive) `g_slist_copy()` here.

```suggestion
	for (GSList *todo = work; todo; todo = todo->next) {
		VipsTile *tile = (VipsTile *) todo->data;
```